dyn_sim/sys/planar/segway.py

```diff
@@ -1 +1,212 @@
+        self._L = L
+        self._J0 = J0
+        self._Km = Km
+        self._R = R
+        self._bt = bt
+        self._l = l
+        self._mass = mass
+
+    def fdyn(self, t: float, s: np.ndarray) -> np.ndarray:
+        """Segway autonomous dynamics.
+
+        Parameters
+        ----------
+        t : float
+            Time. Unused, included for API compliance.
+        s : np.ndarray, shape=(4,)
+            State of segway.
+
+        Returns
+        -------
+        _fdyn : np.ndarray, shape=(4,)
+            Time derivatives of states from autonomous dynamics.
+        """
+        assert s.shape == (4,)
+
+        # states
+        _, phi, dp, dphi = s
+        sinphi = np.sin(phi)
+        cosphi = np.cos(phi)
+
+        # inertia matrix
+        D = np.array(
+            [
+                [self._m0, self._m * self._L * cosphi],
+                [self._m * self._L * cosphi, self._J0],
+            ]
+        )
+        Dinv = np.linalg.inv(D)
+
+        # coriolis & gravity matrix
+        H = np.array(
+            [
+                -self._m * self._L * sinphi * dphi**2
+                + self._bt * (dp - self._R * dphi) / self._R,
+                -self._m * g * self._L * sinphi + self._bt * (dp - self._R * dphi),
+            ]
+        )
+
+        DinvH = Dinv @ H
+        _fdyn = np.array([dp, dphi, -DinvH[0], -DinvH[1]])
+        return _fdyn
+
+    def gdyn(self, t: float, s: np.ndarray) -> np.ndarray:
+        """Segway control dynamics.
+
+        Parameters
+        ----------
+        t : float
+            Time. Unused, included for API compliance.
+        s : np.ndarray, shape=(4,)
+            State of segway.
+
+        Returns
+        -------
+        _gdyn : np.ndarray, shape=(4,)
+            Matrix representing affine control dynamics.
+        """
+        assert s.shape == (4,)
+
+        # states
+        _, phi, _, _ = s
+
+        # sinphi = np.sin(phi) # unused
+        cosphi = np.cos(phi)
+
+        # inertia matrix
+        D = np.array(
+            [
+                [self._m0, self._m * self._L * cosphi],
+                [self._m * self._L * cosphi, self._J0],
+            ]
+        )
+        Dinv = np.linalg.inv(D)
+
+        # input matrix
+        B = np.array([self._Km / self._R, -self._Km])
+
+        DinvB = Dinv @ B
+        _gdyn = np.array([0, 0, DinvB[0], DinvB[1]])
+        return _gdyn
+
+    def A(self, s: np.ndarray, u: np.ndarray) -> np.ndarray:
+        """Linearized autonomous dynamics about (s, u).
+
+        Parameters
+        ----------
+        s : np.ndarray, shape=(4,)
+            State.
+        u : float
+            Input of the segway. Unused, included for API compliance.
+
+        Returns
+        -------
+        _A : np.ndarray, shape=(4, 4)
+            Linearized autonomous dynamics about s.
+        """
+        assert s.shape == (4,)
+        assert u.shape == (1,)
+
+        raise NotImplementedError  # TODO implement linearized dynamics
```

Comment:
after this branch is merged, can you start another PR for adding the segway linearized dynamics? we _could_ also make it part of the ilqr changes, but i'm not so sure that's best... feels a bit dirty to combine them.

alternatively, just implement it with this PR, it would still be self-contained.